backend/app.py

```diff
@@ -76,3 +76,130 @@ def _streamer(last_event_id: int | None = None):
+            """)
+            
+            # Create or replace a function to update the cache
+            cursor.execute("""
+                CREATE OR REPLACE FUNCTION update_stock_price_cache() RETURNS TRIGGER AS $$
+                BEGIN
+                    INSERT INTO stock_price_cache (price_id, payload)
+                    VALUES (NEW.id, (SELECT json_build_object(
+                        'id', NEW.id,
+                        'price', NEW.price,
+                        'created_at', NEW.created_at
+                    )));
+                    RETURN NEW;
+                END;
+                $$ LANGUAGE plpgsql;
+            """)
+            
+            # Create trigger if it doesn't exist (first check if it exists)
+            cursor.execute("""
+                SELECT EXISTS (
+                    SELECT 1 FROM pg_trigger 
+                    WHERE tgname = 'stock_price_cache_trigger'
+                )
+            """)
+            trigger_exists = cursor.fetchone()[0]
+            
+            if not trigger_exists:
+                cursor.execute("""
+                    CREATE TRIGGER stock_price_cache_trigger
+                    AFTER INSERT ON stock_price
+                    FOR EACH ROW
+                    EXECUTE FUNCTION update_stock_price_cache();
+                """)
+            
+            connection.commit()
+            
+            # Function to clean up old cache entries (run periodically)
+            cursor.execute("""
+                CREATE OR REPLACE FUNCTION cleanup_stock_price_cache() RETURNS void AS $$
+                BEGIN
+                    DELETE FROM stock_price_cache 
+                    WHERE created_at < NOW() - INTERVAL '1 day'
+                    AND accessed_at < NOW() - INTERVAL '1 hour';
+                END;
+                $$ LANGUAGE plpgsql;
+            """)
+            
+            # Set up LISTEN/NOTIFY for cache invalidation
+            cursor.execute('LISTEN "stock_price_insert";')
+            connection.commit()
+            
+            # Get initial data from cache if available
+            base_query = """
+                SELECT payload FROM stock_price_cache
+                WHERE price_id > %s
+                ORDER BY price_id ASC
+            """
+            
+            # Use 0 if last_event_id is None
+            cursor.execute(base_query, (last_event_id or 0,))
+            cached_results = cursor.fetchall()
+            
+            # Send cached results immediately
+            for cached_result in cached_results:
+                payload = cached_result[0]
+                price_id = payload['id']
+                
+                # Update access time to indicate this cache entry was used
+                cursor.execute("""
+                    UPDATE stock_price_cache 
+                    SET accessed_at = CURRENT_TIMESTAMP
+                    WHERE price_id = %s
+                """, (price_id,))
+                connection.commit()
```

Comment:
This implementation has a very big disadvantage of repeated commits to the database, which was not the case in the previous version 3.